couchdb-cluster-playground/steps.py

```diff
@@ -0,0 +1,152 @@
+import os
+import subprocess
+import time
+import threading
+
+from couchdbkit import Server
+import requests
+
+import lib.couchdb_custom as couch_info
+
+
+NODES = [
+    {
+        'url': 'http://localhost:5984',
+        'docker_name': 'server-0'
+    },
+    {
+        'url': 'http://localhost:15984',
+        'docker_name': 'server-1'
+    },
+    {
+        'url': 'http://localhost:25984',
+        'docker_name': 'server-2'
+    },
+]
+
+should_create_docs = False  # Boolean used for doc creation thread
+
+
+def create_docs(reporting_interval=4):
+    def create_docs_thead():
+        # thread that runs and inserts documents
+        total_count = 0
+        last_reported_time = 0
+        new_count = 0
+
+        while should_create_docs:
+            doc = {"mydoc": "test"}
+            db.save_doc(doc)
+            total_count += 1
+            new_count += 1
+            cur_time = int(time.time())
+            if ((cur_time - last_reported_time) > reporting_interval):
+                print(f'inserted {total_count} total documents, {new_count} new documents since last reported')
+                new_count = 0
+                last_reported_time = cur_time
+        print(f'insertion done: inserted {total_count} total documents')
+
+    # create database and start thread
+    server = Server("http://admin:password@localhost:5984")
+    db = server.get_or_create_db("shards")
+    x = threading.Thread(target=create_docs_thead)
+    x.start()
+
+
+def node_up(docker_name):
+    # brings up a couchdb node via docker-compose
+    run_command(['docker-compose', 'start', docker_name])
+
+
+def node_down(docker_name):
+    # takes down a couchdb node via docker-compose
+    run_command(['docker-compose', 'stop', docker_name])
+
+
+def run_command(command_list):
+    # runs a command and prints output
+    subprocess.run(command_list, stdout=subprocess.PIPE).stdout.decode('utf-8')
+
+
+def db_is_consistent():
+    # Checks to see if db is consistent based on querying individual nodes and checking doc count.
+    host = 'localhost'
+    port = '5984'
+    local_port = '5984'
+    user = 'admin'
+    password = 'password'
+    with requests.Session() as session:
+        session.auth = (user, password)
+        context = couch_info.SessionContext(host, port, local_port, session)
+        shards = couch_info.get_cluster_shard_details(context, [node['url'] for node in NODES])
+
+        doc_counts = {shard['node']: shard['doc_count'] for shard in shards}
+        print(doc_counts)
+
+        return len(set(doc_counts.values())) == 1
```

Comment:
Would it make sense to also check this against the total number of documents created?

Reply:
Ah currently it doesn't actually check for whether the request succeeded in creating a doc (misnomer in total docs created). I think the shard doc count is the source of truth though? Also, right now multiple runs will go off of the same `shards` db so docs created != shard docs, I thought having a big shard count from previous runs would be helpful as well.